pkg/build/controller/common/util.go

```diff
@@ -42,6 +48,112 @@ func HandleBuildCompletion(build *buildapi.Build, runPolicies []policy.RunPolicy
 	if err := runPolicy.OnComplete(build); err != nil {
 		utilruntime.HandleError(fmt.Errorf("failed to run policy on completed build %s/%s: %v", build.Namespace, build.Name, err))
 	}
+	if err := HandleBuildPruning(build, buildLister, buildConfigLister, buildDeleter); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to prune old builds %s/%s: %v", build.Namespace, build.Name, err))
+	}
+}
+
+// HandleBuildPruning handles the deletion of old successful and failed builds
+// based on settings in the BuildConfig.
+func HandleBuildPruning(build *buildapi.Build, buildLister buildclient.BuildLister, buildConfigLister buildclient.BuildConfigLister, buildDeleter pruner.BuildDeleter) error {
+	if build.Status.Config == nil {
+		glog.V(4).Infof("There is no buildconfig defined for this build: %v/%v", build.Namespace, build.Name)
+		return nil
+	}
+
+	bcList, err := buildConfigLister.List(build.Namespace, kapi.ListOptions{
+		LabelSelector: labels.Set{"app": buildapi.LabelValue(build.Status.Config.Name)}.AsSelector(),
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(bcList.Items) == 0 {
+		glog.V(0).Infof("There are no buildconfigs associated with this build: %v/%v", build.Namespace, build.Name)
+		return nil
+	}
+
+	if bcList.Items[0].Spec.FailedBuildsHistoryLimit != nil || bcList.Items[0].Spec.SuccessfulBuildsHistoryLimit != nil {
+
+		bList, err := buildLister.List(build.Namespace, kapi.ListOptions{
```

Comment:
I'd consider using util.BuildConfigBuilds here, then you may not need your SortBuildsByPhase routine.